packages/gestalt/src/Typeahead.js

```diff
@@ -0,0 +1,351 @@
+    value: string,
+  }) => void => void,
+  onFocus?: () => void,
+  onSelect?: OptionObject => void,
+  value?: string,
+  caret?: boolean,
+  defaultItem?: OptionObject,
+  noResultText?: string,
+  noResultTextColor?: TextColors,
+  resultHeight?: string,
+  hoverColor?: TextColors,
+  textColor?: TextColors,
+  backgroundColor?: TextColors,
+|};
+
+const Typeahead = (props: Props) => {
+  const {
+    accessibilityLabel = 'Demo Search Field',
+    id,
+    onBlur,
+    onChange,
+    onFocus,
+    onSelect,
+    value = '',
+    data,
+    placeholder,
+    size,
+    resultHeight = '50vh',
+    searchField = 'label',
+    defaultItem = null,
+    caret = false,
+    noResultText = 'No Results',
+    noResultTextColor = 'gray',
+    hoverColor = 'lightGray',
+    textColor = 'darkGray',
+    backgroundColor = 'white',
+  } = props;
+
+  // Store original data
+  const dataRef = useRef(data);
+
+  // Utility function for filtering data by value
+  const filterOriginalData = filterValue =>
+    dataRef.current.filter(item =>
+      item[searchField].toLowerCase().includes(filterValue.toLowerCase())
+    );
+
+  // Track input value
+  const [search, setSearch] = useState<string>(value);
+
+  // Track the selected item - could be used to see if someone is selecting the same thing again
+  const [selected, setSelected] = useState<OptionObject | null>(defaultItem);
+
+  const [options, setOptions] = useState<OptionObject[]>(
+    filterOriginalData(search)
+  );
+
+  // Ref to the input
+  const inputRef = useRef();
+
+  // Reference to selected option
+  // let selectedOptionRef;
+  // const getOptionRef = ref => {
+  //   selectedOptionRef = ref;
+  // };
+
+  // Option Container ref
+  // let containerRef;
+  // const getContainerRef = ref => {
+  //   containerRef = ref;
+  // };
+
+  // Handle when input is in and out of focus
+  const componentRef = useRef();
+  const [focused, setFocused] = useState<boolean>(false);
+
+  // When the menu item opens, scroll to selected item
+  // useEffect(() => {
+  //   setTimeout(() => {
+  //     // TODO: Fix scrolling calcultation
+  //     // if (selected !== null && containerRef && selectedOptionRef)
+  //     // eslint-disable-next-line no-use-before-define
+  //     // scrollIntoView(containerRef, selectedOptionRef);
+  //   }, 100);
+
+  //   // eslint-disable-next-line react-hooks/exhaustive-deps
+  // }, [focused, selectedOptionRef]);
+
+  const handleFocus = () => {
+    // Internally set focus status
+    if (componentRef.current)
+      setFocused(componentRef.current.contains(document.activeElement));
+
+    // Run focus callback
+    if (onFocus) onFocus();
+  };
+
+  const handleBlur = () => {
+    // Clear input and reset options
+    if (options.length === 0) {
+      setSearch('');
+      setOptions(dataRef.current);
+    }
+
+    // TODO: Is this the best way to hide the results on blur
+    if (document.activeElement && componentRef.current) {
+      setTimeout(() => {
```

Comment:
@simkessy Using `setTimeout` will break if we unmount the component without clearing the timeout.

Let's figure out a better way to do this.